src/lib/utils/ct_utils.h

```diff
@@ -126,17 +158,23 @@ class Mask final {
       /**
       * Return a Mask<T> which is set if v is == 0 or cleared otherwise
       */
-      static constexpr Mask<T> is_zero(T x) { return Mask<T>(ct_is_zero<T>(x)); }
+      static constexpr Mask<T> is_zero(T x) { return Mask<T>(ct_is_zero<T>(value_barrier<T>(x))); }
 
       /**
       * Return a Mask<T> which is set if x == y
       */
-      static constexpr Mask<T> is_equal(T x, T y) { return Mask<T>::is_zero(static_cast<T>(x ^ y)); }
+      static constexpr Mask<T> is_equal(T x, T y) {
+         const T diff = x ^ y;
+         return Mask<T>::is_zero(value_barrier<T>(diff));
+      }
```

Comment:
I believe it is not enough to apply the `value_barrier` on the result of the xor. If the compiler can somehow infer that both `x` and `y` have a small value-space, it might still come up with some sort of jump table. At this point I'm not sure when an example becomes too contrived, tbh.

Same is true for the other factory methods as well.

Reply:
Unfortunately I think this kind of work is necessarily reactive rather than proactive. The best we can do IMO is dramatically increase the use of constant time tainting which will help detect cases where things are going bad.

That said I do agree is_equal is prone to issues especially since it's not uncommon for one arg to be a constant and the other a secret. I inserted value_barrier on both args